macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+        ''' Extarct the macaroon from a direct successful discharge.
+
+        @param response from direct successful discharge.
+        @return a macaroon object.
+        @raises DischargeError if any error happens.
+        '''
+        try:
+            response_json = response.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access data from discharge endpoint on 200')
+        try:
+            return deserialize(response_json['Macaroon'])
+        except KeyError:
+            raise DischargeError(
+                'Unable to access macaroon from discharge endpoint on 200')
+
+    def _acquire_macaroon_from_wait(self, wait_url):
+        ''' Wait that the user did validate its identity as the get will block
+            until then.
+            If validated then we get the macaroon from the wait endpoint
+            response.
+
+        @param wait_url the get url to call to get a macaroon.
+        @return a macaroon object
+        @raises DischargeError if any error happens.
+        '''
+
+        try:
+            resp = requests.get(wait_url, timeout=self._timeout)
+        except requests.RequestException as exc:
+            raise DischargeError(
+                'Unable to access wait url: {}'.format(exc.message))
+        if not resp.ok:
+            raise DischargeError(
+                'Unexpected status code from wait url: {}'.format(
+                    resp.status_code
+                ))
+        try:
+            response_json = resp.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access json macaroon from wait url'
+            )
+        try:
+            discharge_token = response_json['DischargeToken']
+            macaroon = response_json['Macaroon']
+        except KeyError:
+            raise DischargeError(
+                'Unable to access macaroon or discharge token from'
+                ' discharge endpoint'
+            )
+        if self._discharge_token is None:
+            self._discharge_token = to_base64(discharge_token)
+        return deserialize(macaroon)
+
+    def _extract_urls(self, response):
+        ''' Return the visit and wait URL from response.
+
+        @param response the response from the discharge endpoint.
+        @return the visit and wait URL.
+        @raises DischargeError for ant error during the process response.
+        '''
+        try:
+            response_json = response.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access json from discharge endpoint on 401')
+        try:
+            code = response_json['Code']
+        except KeyError:
+            raise DischargeError(
+                'Unable to access code from discharge endpoint on 401')
+        if code != 'interaction required':
+            raise DischargeError(
+                'Unexpected Code on discharge endpoint: {}'.format(
```

Comment:
It's OK for a discharge endpoint to return a non-interaction-required error. The whole point of the DischargeError type in httpbakery is so that we can find this kind of error. We should return it here too, I think.